### DISCUSSION TITLE
Is it possible to draw networkx edges by attribute?

### DISCUSSION THREAD

**Discussion by adam2392** [networkx/networkx#5637] [General]
Hi, 

I'm trying to extend networkx for usage with causal graphs, which encompass multiple types of edges (i.e. direct and bidirected arrows).

To draw the edges such that they don't overlap I have to indicate the `edge_list` parameter inside networkx. Is it possible to add a parameter to draw edges based on edge attribute?

E.g. this code ends up becoming unwieldy if I have to manually specify every tuple of "direct" and "bidirected" edges.

```
# nx.draw_networkx(dag, pos=pos)
nodes = nx.draw_networkx_nodes(G, pos, node_size=node_sizes, node_color="indigo")
directed_edges = nx.draw_networkx_edges(
    G,
    pos,
    edgelist=[('A', 'B', 'direct')],
    node_size=node_sizes,
    arrowstyle="->",
    arrowsize=10,
    # edge_color=edge_colors,
    edge_cmap=cmap,
    width=2,
    # connectionstyle="arc3,rad=0.1"
)
bd_edges = nx.draw_networkx_edges(
    G,
    pos,
    edgelist=[('A', 'B', 'bidirected')],
    node_size=node_sizes,
    style='dotted',
    # arrowstyle="->",
    arrowsize=10,
    # edge_color=edge_colors,
    edge_cmap=cmap,
    width=2,
    connectionstyle="arc3,rad=0.4"
)

bd_edges = nx.draw_networkx_edges(
    G,
    pos,
    edgelist=[('B', 'A', 'bidirected')],
    node_size=node_sizes,
    style='dotted',
    # arrowstyle="->",
    arrowsize=10,
    # edge_color=edge_colors,
    edge_cmap=cmap,
    width=2,
    connectionstyle="arc3,rad=-0.4"
)
```

**Comment by adam2392** [comment]
If so, I'm happy to submit a PR. 

All I ask is some handy advice on where to start in the code. (Have exp doing open source software, but not familiar with the networkx  codebase yet :p)

**Comment by rossbar** [comment]
> Is it possible to add a parameter to draw edges based on edge attribute?

Currently this is not supported. IIRC this has been discussed before (though I failed to find any relevant links from a quick scan of the issue tracker); extracting the visualization attributes directly from the  graph/node/edge attributes is a significant departure from the way `nx_pylab` currently works. It's an interesting idea (and more akin to how e.g. Graphviz works) but it strikes me as a larger-scale project.

Re: the specific case you mention, I think there are quite a few things you can do to simplify the drawing, e.g. (untested)

```python
nodes = nx.draw_networkx_nodes(G, pos, node_size=node_sizes, node_color="indigo")
# How exactly this is done depends on how you are setting the edge attributes
# For the sake of illustration I assumed you have an "is_bidirectional" edge attr that
# is a bool
d_edges = [(u, v) for u, v, isbd in G.edges(data="is_bidrected") if not isbd]
bd_edges = [(u, v) for u, v, isbd in G.edges(data="is_bidirected") if isbd]

opts = {  # Options common to all edges
    "node_size": node_sizes,
    "arrowsize": 10,
    "edge_cmap": edge_cmap,
    "width": 2,
}

nx.draw_networkx_edges(G, pos, edgelist=d_edges, arrowstyle="->", **opts)
nx.draw_networkx_edges(G, pos, edgelist=bd_edges, arrowstyle="<->", style="dotted", **opts)
```

... give or take an option or two. If you have many graphs that have the same types of attributes, it's generally a good idea to create a function with your preferred visualization params.
